### PR TITLE
Attachment Iteration

### PR DESCRIPTION
**Pull request by dsainati1**
Implements iteration on attachments 
______

- [ ] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter.go [hunk 1]
```diff
@@ -4534,7 +4534,7 @@ func (interpreter *Interpreter) mapMemberValueAuthorization(self Value, memberAc
 }
 
 func (interpreter *Interpreter) getMemberWithAuthMapping(self Value, locationRange LocationRange, identifier string) Value {
-	result := interpreter.getMember(self, locationRange, identifier)
+	result := interpreter.getMember(self, locationRange, identifier, nil)
```
Comment (SupunS) [inline runtime/interpreter/interpreter.go:4537]: Should we pass `UnauthorizedAccess` here instead of `nil`?

File: runtime/interpreter/value.go [hunk 2]
```diff
@@ -15157,6 +15161,51 @@ func (v *CompositeValue) GetAttachments(interpreter *Interpreter, locationRange
 	return attachments
 }
 
+func (v *CompositeValue) forEachAttachmentFunction(interpreter *Interpreter, baseAuthorization Authorization, locationRange LocationRange) Value {
```
Comment (SupunS) [inline runtime/interpreter/value.go:15164]: `baseAuthorization` seems to be unused here. Is it needed for any follow-up PRs?

Comment (dsainati1) [reply runtime/interpreter/value.go:15164]: Nope, that's a leftover from before when entitlements had runtime behavior. Wish Go's unused variable lints also applied to named function parameters.

File: runtime/interpreter/interpreter.go (hunk 1 above)
Comment (dsainati1) [reply runtime/interpreter/interpreter.go:4537]: This argument isn't actually needed anymore (see comment below), so I just removed it.

**Review by SupunS — APPROVED** [review]

File: runtime/interpreter/value.go [hunk 3]
```diff
@@ -15304,6 +15308,51 @@ func (v *CompositeValue) GetAttachments(interpreter *Interpreter, locationRange
...
+				attachmentReference := NewEphemeralReferenceValue(
+					interpreter,
+					attachmentReferenceAuth,
+					attachment,
+					attachmentType,
+				)
+
+				invocation := NewInvocation(
+					interpreter,
+					nil,
```
[the remark this answers is not in the document]
Comment (dsainati1) [reply runtime/interpreter/value.go:15339]: No, we don't want `self` to be visible in the body of this function unless it's already visible. I.e. if you call this outside of a composite, we shouldn't make `self` available inside this invocation.

File: runtime/sema/type.go [hunk 4]
```diff
@@ -4205,6 +4205,45 @@ func (t *CompositeType) IsValidIndexingType(ty Type) bool {
...
+									},
+								),
+							},
+						},
+						ReturnTypeAnnotation: NewTypeAnnotation(VoidType),
+					},
+				),
+			},
+		},
+		ReturnTypeAnnotation: NewTypeAnnotation(VoidType),
```
Comment (turbolent) [inline runtime/sema/type.go:4243]:
```suggestion
						ReturnTypeAnnotation: VoidTypeAnnotation,
					},
				),
			},
		},
		ReturnTypeAnnotation: VoidTypeAnnotation,
```

File: runtime/tests/interpreter/attachments_test.go [hunk 5]
```diff
@@ -1947,3 +1947,397 @@ func TestInterpretAttachmentDefensiveCheck(t *testing.T) {
...
+            attachment B for S {
+                fun foo(): Int { return 10 }
+            }
+            attachment C for S {
+                fun foo(_ x: Int): Int { return 8 + x }
+            }
+            fun test(): Int {
+                var s = attach C() to attach B() to attach A() to S()
+                var i = 0
+                s.forEachAttachment(fun(attachment: &AnyStructAttachment) {
```
Comment (turbolent) [inline runtime/tests/interpreter/attachments_test.go:2024]: Nice!

File: runtime/interpreter/value.go [hunk 6]
```diff
@@ -15304,6 +15308,51 @@ func (v *CompositeValue) GetAttachments(interpreter *Interpreter, locationRange
...
+					[]Value{attachmentReference},
+					[]sema.Type{sema.NewReferenceType(interpreter, attachmentType, sema.UnauthorizedAccess)},
+					nil,
+					locationRange,
+				)
+				functionValue.invoke(invocation)
+			}
+
+			v.forEachAttachment(interpreter, locationRange, fn)
```
Comment (turbolent) [inline runtime/interpreter/value.go:15350]: Do we also need mutation prevention during iteration here?

**Review by turbolent — COMMENTED** [review]
Nice!

File: runtime/interpreter/value.go (hunk 6 above)
Comment (dsainati1) [reply runtime/interpreter/value.go:15350]: Yeah, the `forEachAttachment` function handles the state here, and then during adding and removing of attachments we check for the state (see `VisitAttachExpression` and `VisitRemoveStatement`) respectively.

**Review by turbolent — APPROVED** [review]
Great to see that we can finally enable this feature! 👏 🎉

**Pull request merged**